tests/unit/browser/webkit/network/test_pac.py

```diff
@@ -0,0 +1,242 @@
+
+from qutebrowser.browser.webkit.network import pac
+
+
+pytestmark = pytest.mark.usefixtures('qapp')
+
+
+def _pac_common_test(test_str, use_fixed_dt=True):
+    date = QDate(2016, 8, 26)
+    time = QTime(10, 30, 0)
+    tz = 7200
+
+    fun_str_f = """
+        function FindProxyForURL(domain, host) {{
+            {}
+            return "DIRECT; PROXY 127.0.0.1:8080; SOCKS 192.168.1.1:4444";
+        }}
+    """
+
+    fun_str = fun_str_f.format(test_str)
+
+    if use_fixed_dt:
+        dt = QDateTime(date, time, Qt.OffsetFromUTC, tz)
+        res = pac.PACResolver(fun_str, fixed_dt=dt)
+    else:
+        res = pac.PACResolver(fun_str)
+
+    proxies = res.resolve(QNetworkProxyQuery(QUrl("https://example.com/test")))
+    assert len(proxies) == 3
+    assert proxies[0].type() == QNetworkProxy.NoProxy
+    assert proxies[1].type() == QNetworkProxy.HttpProxy
+    assert proxies[1].hostName() == "127.0.0.1"
+    assert proxies[1].port() == 8080
+    assert proxies[2].type() == QNetworkProxy.Socks5Proxy
+    assert proxies[2].hostName() == "192.168.1.1"
+    assert proxies[2].port() == 4444
+
+
+def _pac_equality_test(call, expected):
+    test_str_f = """
+        var res = ({0});
+        var expected = ({1});
+        if(res !== expected) {{
+            throw new Error("failed test {0}: got '" + res + "', expected '" + expected + "'");
+        }}
+    """
+    _pac_common_test(test_str_f.format(call, expected))
+
+
+def _pac_except_test(caplog, call):
+    test_str_f = """
+        var thrown = false;
+        try {{
+            var res = ({0});
+        }} catch(e) {{
+            thrown = true;
+        }}
+        if(!thrown) {{
+            throw new Error("failed test {0}: got '" + res + "', expected exception");
+        }}
+    """
+    with caplog.at_level(logging.ERROR):
+        _pac_common_test(test_str_f.format(call))
+
+
+def _pac_noexcept_test(call):
+    test_str_f = """
+        var res = ({0});
+    """
+    _pac_common_test(test_str_f.format(call), use_fixed_dt=False)
+
+
+# pylint: disable=line-too-long, invalid-name
+
+
+def test_isPlainHostName():
+    _pac_equality_test("isPlainHostName('www')", "true")
+    _pac_equality_test("isPlainHostName('www.netscape.com')", "false")
+
+
+def test_dnsDomainIs():
+    _pac_equality_test("dnsDomainIs('www.netscape.com', '.netscape.com')", "true")
+    _pac_equality_test("dnsDomainIs('www', '.netscape.com')", "false")
+    _pac_equality_test("dnsDomainIs('www.mcom.com', '.netscape.com')", "false")
+
+
+def test_localHostOrDomainIs():
+    _pac_equality_test("localHostOrDomainIs('www.netscape.com', 'www.netscape.com')", "true")
+    _pac_equality_test("localHostOrDomainIs('www', 'www.netscape.com')", "true")
+    _pac_equality_test("localHostOrDomainIs('www.mcom.com', 'www.netscape.com')", "false")
+    _pac_equality_test("localHostOrDomainIs('home.netscape.com', 'www.netscape.com')", "false")
+
+
+def test_isResolvable():
```

Comment:
We shouldn't do network requests in tests (for various reasons), so this should probably [monkeypatch](http://doc.pytest.org/en/latest/monkeypatch.html) `QHostInfo` to avoid that.